### PR TITLE
issue-21715 fetching metric values with composite(all) dimension keys

### LINKED ISSUE #21715: AzureMonitor Receiver does not fetch combinational dimensions of metric values.
### Component(s)

receiver/azuremonitor

### Describe the issue you're reporting

If metric has multiple dimensions, metricValues are fetched by single dimension instead of combining all dimensions.  With single dimension metric value is average value of other dimensions. 

In below example without all dimensions in the filter, 'metadata_nodepool' provides average value of 2 'metadata_node' datapoints. But cannot combine dimensions of metadata_node under  'metadata_nodepool' dimensions.

```
Metric #44
Descriptor:
     -> Name: azure_node_cpu_usage_millicores_total
NumberDataPoints #0
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_nodepool: Str(general1)
Value: 625.500000

NumberDataPoints #1
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss00006x)
Value: 506.000000

NumberDataPoints #2
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss000062)
Value: 745.000000
```

With all Dimensions applied:
 Here each datapoint provides all related dimension values. With this information, it is easy to apply multiple filters.

```
Metric #44
Descriptor:
     -> Name: azure_node_cpu_usage_millicores_total

NumberDataPoints #0
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss00006x)
     -> metadata_nodepool: Str(general1)
Value: 506.000000

NumberDataPoints #1
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss000062)
     -> metadata_nodepool: Str(general1)
Value: 745.000000

```

In Azuremonitor receiver, [getResourceMetricsValuesRequestOptions](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/azuremonitorreceiver/scraper.go#L372-L374) func, sinlge dimension key alone provided. Instead all possible dimension keys for the metric have to be given. 

Sample MetricClient REST Request with all dimensions.

`https://management.azure.com/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue/providers/Microsoft.Insights/metrics?timespan=2023-04-29T02:20:00Z/2023-05-04T04:20:00Z&interval=PT1H&api-version=2018-01-01&metricnamespace=Microsoft.ContainerService/managedClusters&metricnames=node_cpu_usage_millicores&$filter=nodepool eq '*' and node eq '*'`

### PR DESCRIPTION
**Pull request by rajim17**
**Description:**

If metric has multiple dimensions, metricValues are fetched by single dimension instead of combining all dimensions. With single dimension metric value is average value of other dimensions.

In below example without all dimensions in the filter, 'metadata_nodepool' provides average value of 2 'metadata_node' datapoints. But cannot combine dimensions of metadata_node under 'metadata_nodepool' dimensions.

```
Metric #44
Descriptor:
     -> Name: azure_node_cpu_usage_millicores_total
NumberDataPoints #0
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_nodepool: Str(general1)
Value: 625.500000

NumberDataPoints #1
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss00006x)
Value: 506.000000

NumberDataPoints #2
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss000062)
Value: 745.000000
```
With all Dimensions applied:
Here each datapoint provides all related dimension values. With this information, it is easy to apply multiple filters.

```
Metric #44
Descriptor:
     -> Name: azure_node_cpu_usage_millicores_total

NumberDataPoints #0
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss00006x)
     -> metadata_nodepool: Str(general1)
Value: 506.000000

NumberDataPoints #1
Data point attributes:
     -> azuremonitor.resource_id: Str(/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue)
     -> metadata_node: Str(aks-general1-vmss000062)
     -> metadata_nodepool: Str(general1)
Value: 745.000000
```

In Azuremonitor receiver, [getResourceMetricsValuesRequestOptions](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/azuremonitorreceiver/scraper.go#L372-L374) func, single dimension key alone provided. Instead all possible dimension keys for the metric have to be given. 
Solution:
    with metricresource definitions,  Generate hash string with all dimensions and store the metrics based on hash instead of single dimension + timegrain filter.

Sample MetricClient REST Request with all dimensions in filter:

`https://management.azure.com/subscriptions/test/resourceGroups/aks-dev2-blue/providers/Microsoft.ContainerService/managedClusters/aks-dev2-blue/providers/Microsoft.Insights/metrics?timespan=2023-04-29T02:20:00Z/2023-05-04T04:20:00Z&interval=PT1H&api-version=2018-01-01&metricnamespace=Microsoft.ContainerService/managedClusters&metricnames=node_cpu_usage_millicores&$filter=nodepool eq '*' and node eq '*'`

Fixes #21715

**Testing:** Testdata with multiple dimensions are working.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by rajim17** [comment]
@altuner @codeboten - Could you help review the changes ?

**Review by altuner — CHANGES_REQUESTED** [review]
left small comments

File: receiver/azuremonitorreceiver/scraper.go [hunk 1]
```diff
@@ -64,18 +66,26 @@ const (
...
 
 type azureResource struct {
-	metricsByCompositeKey     map[metricsCompositeKey]*azureResourceMetrics
+	metricsByCompositeKey     map[string]*azureResourceMetrics
 	metricsDefinitionsUpdated time.Time
 	tags                      map[string]*string
 	location                  string
 }
 
 type metricsCompositeKey struct {
```
[the remark this answers is not in the document]
Comment (rajim17) [reply receiver/azuremonitorreceiver/scraper.go]: Yes. I removed metricsCompositeKey struct and directly injecting into azureResourceMetrics.

**Review by altuner — APPROVED** [review]
LGTM

**Review by altuner — CHANGES_REQUESTED** [review]

File: receiver/azuremonitorreceiver/scraper.go [hunk 2]
```diff
@@ -64,18 +66,15 @@ const (
 )
 
 type azureResource struct {
-	metricsByCompositeKey     map[metricsCompositeKey]*azureResourceMetrics
+	metricsByCompositeKey     map[string]*azureResourceMetrics
```
[the remark this answers is not in the document]
Comment (rajim17) [reply receiver/azuremonitorreceiver/scraper.go:69]: @altuner - dimensions is slice in type metricsCompositeKey hence can't be used as a key in map. Have to use hash string as key. I can revert to previous commit keeping metricsCompositeKey and still using it's hash string as key to map.

[the remark this answers is not in the document]
Comment (rajim17) [reply receiver/azuremonitorreceiver/scraper.go:69]: I'm not sure of any other way other than hashing to store the key to map. Yes, I should add sort to make sure dimension strings are ordered.  Let me add that. Do you think metricsCompositeKey struct can be useful ?

[the remark this answers is not in the document]
Comment (rajim17) [reply receiver/azuremonitorreceiver/scraper.go]: that is a very good idea. I just changed to comma separated strings and used compositeKey as key in the map.

**Comment by rajim17** [comment]
> @rajim17 please fix the breaking test

Integration_test on rabbitmqreceiver  was failing. I triggered the build process with dummy commit message and tests are running. I will check once build is complete.

**Comment by rajim17** [comment]
@altuner - All tests are passed.

**Review by altuner — APPROVED** [review]

**Comment by rajim17** [comment]
@codeboten - Could you help reviewing the PR ?

**Comment by junhuangli** [comment]
@altuner We are waiting for this fix. Could you merge this pull request or is there anything still pending?

**Comment by rajim17** [comment]
@codeboten - Could you help merging this PR ? It is already approved by code owner @altuner .

**Comment by junhuangli** [comment]
@codeboten Could you take another look at this PR :)

**Review by codeboten — APPROVED** [review]

**Pull request merged**